### PR TITLE
docs: how to reproduce released artifacts

### PR DESCRIPTION
**Pull request by burgerdev**
### Context

There's an implicit assumption that our release artifacts are reproducible, but the exact requirements are unclear. According to reproducible-builds.org, a reproducible build needs to come with a build environment definition.

A user found an issue where the local Nix store state influences build outcomes. This was not caught by our tests because they are only testing a narrow build environment.

### Proposed change(s)

* Fix the issue and add a test for it.
* Document our reproducibility claims, in particular that
  - Artifacts are expected to be reproducible (and it's a bug if not).
  - Set a baseline for reasonable host environments.
  - Our supported and tested environment for reproduction is defined by the workflow.

### Additional info

- [AB#4797](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/4797)
- #3453

### Checklist

- [x] Run the E2E tests that are relevant to this PR's changes
  - [x] [successful repro](https://github.com/edgelesssys/constellation/actions/runs/11457947389)
  - [x] [failed repro without `--hard-deference`](https://github.com/edgelesssys/constellation/actions/runs/11455789295)
  - [x] [after upgrade to 24.04](https://github.com/edgelesssys/constellation/actions/runs/11499461551)
- [x] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

File: docs/docs/workflows/reproducible-builds.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+   //image/system:qemu_qemu-vtpm_stable
+   ```
+
+5. Compare the build result with the downloaded release artifact.
+
+## Feedback
+
+Reproduction failures often indicate a bug in the build system or in the build definitions.
+Therefore, we're interested in any reproducibility issues you might encounter.
+[Start a bug report](https://github.com/edgelesssys/constellation/issues/new/choose) and describe the details of your build.
```
Comment (3u13r) [inline docs/docs/workflows/reproducible-builds.md:67]: Is it too much to ask the user to give us (at least a snippet) of the diffoscope output? Then we already have a clue where to look.

**Review by elchead — APPROVED** [review]

File: docs/docs/workflows/reproducible-builds.md (hunk 1 above)
Comment (burgerdev) [reply docs/docs/workflows/reproducible-builds.md:67]: No, let me add that (and/or the need for the binary itself).

File: .github/workflows/reproducible-builds.yml [hunk 2]
```diff
@@ -24,19 +25,38 @@ jobs:
...
         uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
         with:
           ref: ${{ !github.event.pull_request.head.repo.fork && github.head_ref || '' }}
 
-      - name: Setup bazel
+      - name: Setup dependencies
         uses: ./.github/actions/setup_bazel_nix
+        if: matrix.deps == 'conventional'
+
+      - name: Setup Bazel (eccentric)
```
[the remark this answers is not in the document]
Comment (burgerdev) [reply .github/workflows/reproducible-builds.yml:48]: I wonder if I should perform the nix installation manually, too, to serve as build environment setup instructions.

**Review by daniel-weisse — APPROVED** [review]

File: .github/workflows/reproducible-builds.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (burgerdev) [reply .github/workflows/reproducible-builds.yml]: Removed the action and added a manual download/exec of the nix-installer. Renovate should upgrade it every once in a while with the config change.

File: .github/workflows/reproducible-builds.yml [hunk 3]
```diff
@@ -24,19 +25,38 @@ jobs:
             - "cli_enterprise_linux_amd64"
             - "cli_enterprise_linux_arm64"
             - "cli_enterprise_windows_amd64"
-          runner: ["ubuntu-22.04", "ubuntu-20.04"]
+          runner:
+            - "ubuntu-22.04"
+            - "ubuntu-20.04"
```
[the remark this answers is not in the document]
Comment (burgerdev) [reply .github/workflows/reproducible-builds.yml]: Done

File: docs/docs/workflows/reproducible-builds.md (hunk 1 above)
Comment (burgerdev) [reply docs/docs/workflows/reproducible-builds.md]: Done

File: .github/workflows/reproducible-builds.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (burgerdev) [reply .github/workflows/reproducible-builds.yml]: Unfortunately, the image build does not seem to work on 24.04 - still trying to understand why.

**Review by m1ghtym0 — APPROVED** [review]
Awesome, thanks!

File: docs/docs/workflows/reproducible-builds.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# Reproduce released artifacts
+
+Constellation has first-class support for [Reproducible Builds](https://reproducible-builds.org).
```
Comment (thomasten) [inline docs/docs/workflows/reproducible-builds.md:3]:
```suggestion
Constellation has first-class support for [reproducible builds](https://reproducible-builds.org).
```
nit: you linked the Reproducible Builds project, but the sentence refers to the concept of reproducible builds, right?

File: docs/docs/workflows/reproducible-builds.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+## Run the build
+
+The following instructions outline qualitatively how to reproduce a build.
+Constellation implements these instructions in the [Reproducible Builds workflow](https://github.com/edgelesssys/constellation/actions/workflows/reproducible-builds.yml), which continuously tests for reproducibility.
+The workflow is a good place to look up specific version numbers and build steps.
+
+1. Check out the Constellation repository at the tag corresponding to the release.
+
+   ```sh
```
Comment (thomasten) [inline docs/docs/workflows/reproducible-builds.md:35]:
```suggestion
   ```bash
```
nit: I think we have highlighting for `bash`, but not `sh` in our docusaurus instance

File: docs/docs/workflows/reproducible-builds.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+   //image/system:aws_aws-sev-snp_stable
+   //image/system:azure_azure-sev-snp_stable
+   //image/system:azure_azure-tdx_stable
+   //image/system:gcp_gcp-sev-es_stable
+   //image/system:gcp_gcp-sev-snp_stable
+   //image/system:openstack_qemu-vtpm_stable
+   //image/system:qemu_qemu-vtpm_stable
+   ```
+
+5. Compare the build result with the downloaded release artifact.
```
Comment (thomasten) [inline docs/docs/workflows/reproducible-builds.md:61]: Is this also valid for images or do we need to explain more here?

Comment (burgerdev) [reply docs/docs/workflows/reproducible-builds.md:61]: We don't publish images in a location that is easily downloadable, do we? In any case, one would probably want to verify the PCR content corresponding to the image.

Comment (burgerdev) [reply docs/docs/workflows/reproducible-builds.md:54]: I removed images from the document for now - coming up with a story for verifying them is a bit out of scope for this fix.

**Review by thomasten — APPROVED** [review]

**Review by 3u13r — APPROVED** [review]
LGTM

**Pull request merged**